test/unittest/datasets/unittest_random_data_producers.cpp

```diff
@@ -0,0 +1,66 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file data_producer_common_tests.cpp
+ * @date 12 July 2021
+ * @brief Common test for nntrainer dataset producers (Param Tests)
+ * @see	https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug No known bugs except for NYI items
+ */
+#include <gtest/gtest.h>
+
+#include <algorithm>
+
+#include <data_producer_common_tests.h>
+#include <random_data_producers.h>
+
+DataProducerValidatorType random_onehot_validator(float min, float max) {
+  /// input validator: every value is in range of min, max
+  auto input_valid = [min, max](const nntrainer::Tensor &t) {
+    auto data = t.getData();
+    for (unsigned int i = 0; i < t.length(); ++i) {
+      if (*data < min || max < *data) {
+        return false;
+      }
+      data++;
+    }
+    return true;
+  };
+
+  /// label validator: sum of all is equal to batch
+  auto label_valid = [](const nntrainer::Tensor &t) {
+    return fabs(t.batch() - t.sum({0, 1, 2, 3}).getValue(0, 0, 0, 0) < 1e-7);
+  };
+
+  auto f = [input_valid,
+            label_valid](const std::vector<nntrainer::Tensor> &inputs,
+                         const std::vector<nntrainer::Tensor> &labels) -> bool {
+    bool is_inputs_valid =
+      std::all_of(inputs.begin(), inputs.end(), input_valid);
+    bool is_labels_valid =
+      std::all_of(labels.begin(), labels.end(), label_valid);
+    return is_inputs_valid && is_labels_valid;
+  };
+
+  return f;
+}
+auto random_onehot_success = DataProducerSemanticsParamType(
+  createDataProducer<nntrainer::RandomDataOneHotProducer>,
+  {"min=0", "max=1", "size=10"}, {{3, 2, 4, 5}}, {{3, 1, 1, 10}},
+  random_onehot_validator(0, 1), DataProducerSemanticsExpectedResult::SUCCESS);
+
+auto random_onehot_min_over_max = DataProducerSemanticsParamType(
+  createDataProducer<nntrainer::RandomDataOneHotProducer>,
+  {"min=2", "max=1", "size=10"}, {{3, 2, 4, 5}}, {{3, 1, 1, 10}}, nullptr,
```

Comment:
is this testing the random_data_producer or testing the test setup itself :)?

Reply:
It's kinda both :p

Reply:
I think this test is a bit misleading, and this type of tests exists in future PR as well. The testing of the test setup can be done at once, and no need for the data producer types to do it again.

Reply:
Could you explain why it's misleading? This test case tests random producer.

I mean I have tested the test setup while building this test, and if test code runs fine, it means test setup is correct and the test is correct so it's testing both.

Reply:
> Could you explain why it's misleading? This test case tests random producer.

As per the test, this test will check if the validator if nullptr and then the other tests performed are exactly the same as the tests above it. So, in this way, I couldn't see which extra test does this unit test do.

Reply:
the above test is positive test (DataProducerSemanticsExpectedResult::SUCCESS)
while this test is negative test (DataProducerSemanticsExpectedResult::FAIL_AT_FINALIZE), it is negative test because the case is that the property value "min" is over "max" which is meant to fail, it is not failing because the validator is nullptr but finalize fails.  I am not getting why it's exactly the same as the test above it. Wonder if I am missing something.

Reply:
aah, I thought it was because the validator is nullptr.
then its fine.